flake:  TestExpMcpReporter

Seen [here](https://productionresultssa9.blob.core.windows.net/actions-results/ec7003ab-c9ff-4c11-92cc-3dd0f7936985/workflow-job-run-ab72ea0f-a08e-50ed-9de2-54b022c031cd/logs/job/job-logs.txt?rsct=text%2Fplain&se=2025-06-18T17%3A41%3A50Z&sig=FhT50BU5tH4xaHPjKFIlimw%2FqX5fAof80VRJsJFOdxU%3D&ske=2025-06-19T05%3A24%3A10Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2025-06-18T17%3A24%3A10Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2025-05-05&sp=r&spr=https&sr=b&st=2025-06-18T17%3A31%3A45Z&sv=2025-05-05)

```
2025-06-18T12:54:10.6485918Z     exp_mcp_test.go:867: 
2025-06-18T12:54:10.6486341Z         	Error Trace:	/home/runner/work/coder/coder/cli/exp_mcp_test.go:867
2025-06-18T12:54:10.6487073Z         	            				/opt/hostedtoolcache/go/1.24.4/x64/src/runtime/asm_amd64.s:1700
2025-06-18T12:54:10.6487510Z         	Error:      	Received unexpected error:
2025-06-18T12:54:10.6488200Z         	            	running command "coder exp mcp server": read /dev/pts/0: file already closed
2025-06-18T12:54:10.6488625Z         	Test:       	TestExpMcpReporter/OK
```

Seen locally

```
=== FAIL: cli TestExpMcpReporter/OK (0.69s)
    coderdtest.go:596: coderdtest server listening on 127.0.0.1:45103
    coderdtest.go:596: coderdtest server started on http://127.0.0.1:45103
    t.go:106: 2025-06-24 10:54:57.431 [debu]  coderd.metrics_cache: metrics refreshed  name="deployment stats"  interval=5m0s  took=7.053µs
    t.go:106: 2025-06-24 10:54:57.431 [debu]  coderd.dbrollup: rolling up data
    t.go:106: 2025-06-24 10:54:57.431 [debu]  coderd.dbrollup: rolled up data  took=430.414µs  event={"template_usage_stats":true}
    t.go:106: 2025-06-24 10:54:57.431 [debu]  coderd.dbrollup: next rollup at  next="2025-06-24T10:55:00Z"
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: no valid keys detected, inserting new key  feature=tailnet_resume
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: inserted new key for feature  feature=tailnet_resume
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: no valid keys detected, inserting new key  feature=workspace_apps_token
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: inserted new key for feature  feature=workspace_apps_token
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: no valid keys detected, inserting new key  feature=workspace_apps_api_key
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: inserted new key for feature  feature=workspace_apps_api_key
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: no valid keys detected, inserting new key  feature=oidc_convert
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.keyrotator: inserted new key for feature  feature=oidc_convert
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.acquirer: subscribed to job postings
    t.go:106: 2025-06-24 10:54:57.446 [warn]  coderd.site: could not parse install.sh, it will be unavailable  error="open install.sh: file does not exist"
    t.go:106: 2025-06-24 10:54:57.446 [debu]  coderd.servertailnet.net.wgengine: [v1] using fake (no-op) tun device
    t.go:106: 2025-06-24 10:54:57.446 [debu]  coderd.servertailnet.net.wgengine: [v1] using fake (no-op) OS network configurator
    t.go:106: 2025-06-24 10:54:57.446 [debu]  coderd.servertailnet.net.wgengine: [v1] using fake (no-op) DNS configurator
    t.go:106: 2025-06-24 10:54:57.436 [debu]  coderd.metrics_cache: metrics refreshed  name="template build times"  interval=1h0m0s  took=5.11976ms
    t.go:106: 2025-06-24 10:54:57.444 [debu]  coderd.keyrotator: ctx canceled, stopping key rotation
    t.go:106: 2025-06-24 10:54:57.454 [debu]  coderd.servertailnet.net.wgengine: dns: using dns.noopManager
    t.go:106: 2025-06-24 10:54:57.454 [debu]  coderd.servertailnet.net.wgengine: link state: interfaces.State{defaultRoute=eth0 ifs={docker0:[172.17.0.1/16] eth0:[172.18.0.4/16]} v4=true v6=false}
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: [warning] failed to force-set UDP read buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: [warning] failed to force-set UDP write buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: dontfrag: success on [::]:53300
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: [warning] failed to force-set UDP read buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: [warning] failed to force-set UDP write buffer size to 7340032: operation not permitted; using kernel default values (impacts throughput only)
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: dontfrag: success on 0.0.0.0:51004
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: [v1] couldn't create raw v4 disco listener, using regular listener instead: raw disco listening disabled, SO_MARK unavailable
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: [v1] couldn't create raw v6 disco listener, using regular listener instead: raw disco listening disabled, SO_MARK unavailable
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: disco key = d:e8291fe57a7dc5e6
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: Creating WireGuard device...
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: Bringing WireGuard device up...
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] UDP bind has been updated
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Interface state was Down, requested Up, now Up
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: Bringing router up...
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: [v1] warning: fakeRouter.Up: not implemented.
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: Clearing router settings...
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: [v1] warning: fakeRouter.Set: not implemented.
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: Starting network monitor...
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: Engine created.
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet: magicsock debug logging disabled, use CODER_MAGICSOCK_DEBUG_LOGGING=true to enable
    t.go:106: 2025-06-24 10:54:57.455 [debu]  coderd.servertailnet.net.wgengine: magicsock: SetPrivateKey called (init)
    t.go:106: 2025-06-24 10:54:57.467 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Routine: receive incoming mkReceiveFunc - started
    t.go:106: 2025-06-24 10:54:57.468 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Routine: receive incoming mkReceiveFunc - started
    t.go:106: 2025-06-24 10:54:57.468 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Routine: receive incoming receiveDERP - started
    t.go:106: 2025-06-24 10:54:57.482 [debu]  coderd.servertailnet: updating engine network map  network_map="netmap: self: [ZXp0Z] auth=machine-unknown u=? [fd7a:115c:a1e0:4e0d:9f95:1a52:b21:4fb6/128]\n"
    t.go:106: 2025-06-24 10:54:57.482 [debu]  coderd.servertailnet: skipped sending node; no node callback
    t.go:106: 2025-06-24 10:54:57.487 [debu]  coderd.coord: initPeer  peer_id=eb5f0702-71b0-4bc2-821c-6ce9bc88a40f  peer_name=inmemoryeb5f0702-71b0-4bc2-821c-6ce9bc88a40f
    t.go:106: 2025-06-24 10:54:57.536 [debu]  coderd.coord: initial tunnel peers  peer_id=eb5f0702-71b0-4bc2-821c-6ce9bc88a40f  peer_name=inmemoryeb5f0702-71b0-4bc2-821c-6ce9bc88a40f  tunnel_peers=[]
    t.go:106: 2025-06-24 10:54:57.487 [debu]  coderd.workspaceapps.stats_collector: workspace app stats collector started
    t.go:106: 2025-06-24 10:54:57.502 [debu]  coderd: entitlements will not be refreshed  request_id=c837253c-de19-4cee-8cf8-6200879457a5
    t.go:106: 2025-06-24 10:54:57.519 [debu]  coderd.servertailnet.net.wgengine: [v1] magicsock: got updated network map; 0 peers
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.netstack: [v2] netstack: registered IP fd7a:115c:a1e0:4e0d:9f95:1a52:b21:4fb6/128
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: wgengine: Reconfig: configuring userspace WireGuard config (with 0/0 peers)
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] UAPI: Updating private key
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: wgengine: Reconfig: configuring router
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: [v1] warning: fakeRouter.Set: not implemented.
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: wgengine: Reconfig: configuring DNS
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: dns: Set: {DefaultResolvers:[] Routes:{} SearchDomains:[] Hosts:0}
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: dns: Resolvercfg: {Routes:{} Hosts:0 LocalDomains:[]}
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: dns: OScfg: {Nameservers:[] SearchDomains:[] MatchDomains:[] Hosts:[]}
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet.net.wgengine: [v1] wgengine: Reconfig done
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd.servertailnet: updating engine filter  filter="&{logf:0x223dec0 local:0xc06c73d9b0 logIPs:0xc06c73d9c8 matches4:[{IPProto:[6 17 1 58 132] Srcs:[{ip:{addr:{hi:0 lo:281470681743360} z:{value:0xc000011728}} bitsPlusOne:1}] Dsts:[{Net:{ip:{addr:{hi:0 lo:281470681743360} z:{value:0xc000011728}} bitsPlusOne:1} Ports:{First:0 Last:65535}}] Caps:[]}] matches6:[{IPProto:[6 17 1 58 132] Srcs:[{ip:{addr:{hi:0 lo:0} z:{value:0xc000011740}} bitsPlusOne:1}] Dsts:[{Net:{ip:{addr:{hi:0 lo:0} z:{value:0xc000011740}} bitsPlusOne:1} Ports:{First:0 Last:65535}}] Caps:[]}] cap4:[] cap6:[] state:0xc0d8644760 shieldsUp:false}"
    t.go:106: 2025-06-24 10:54:57.567 [debu]  coderd: POST  host=localhost:45103  path=/api/v2/users/first  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.471698947Z"  requestor_name=System  took=96.033864ms  status_code=201  latency_ms=96  request_id=c837253c-de19-4cee-8cf8-6200879457a5
    t.go:106: 2025-06-24 10:54:57.571 [debu]  coderd.servertailnet: obtained tailnet API v2+ client
    t.go:106: 2025-06-24 10:54:57.577 [debu]  coderd.servertailnet: serving coordinator
    t.go:106: 2025-06-24 10:54:57.577 [debu]  coderd.servertailnet: skipped sending node; no PreferredDERP  node="&{ID:nodeid:72816b41754b5e5f AsOf:2025-06-24 10:54:57.57765 +0000 UTC Key:nodekey:657a74666d19729529d8438b3a5b04020e9b9d087d700088eb8811518973ed3a DiscoKey:discokey:e8291fe57a7dc5e6824ed8f3ab5e5a5a8388621de4532f1f0260246a4a39e26d PreferredDERP:0 DERPLatency:map[] DERPForcedWebsocket:map[] Addresses:[fd7a:115c:a1e0:4e0d:9f95:1a52:b21:4fb6/128] AllowedIPs:[fd7a:115c:a1e0:4e0d:9f95:1a52:b21:4fb6/128] Endpoints:[]}"
    t.go:106: 2025-06-24 10:54:57.583 [debu]  coderd: POST  host=localhost:45103  path=/api/v2/users/login  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.582062885Z"  requestor_id=402d5ce9-49bc-4df5-a39d-35ffb6d3cbb8  requestor_name=testuser  requestor_email=""  took=946.457µs  status_code=201  latency_ms=0  request_id=64799755-a38d-479d-9f21-eea2b04cf09d
    t.go:106: 2025-06-24 10:54:57.691 [debu]  coderd: POST  host=localhost:45103  path=/api/v2/users  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.615781652Z"  requestor_id=402d5ce9-49bc-4df5-a39d-35ffb6d3cbb8  requestor_name=testuser  requestor_email=""  took=76.017767ms  status_code=201  latency_ms=76  request_id=590b7865-200d-4272-8242-709b72c79bc9
    t.go:106: 2025-06-24 10:54:57.696 [debu]  coderd: POST  host=localhost:45103  path=/api/v2/users/login  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.695477316Z"  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=1.057278ms  status_code=201  latency_ms=1  request_id=6be58cdb-35ce-4a52-ae9b-a9516381acb2
    t.go:106: 2025-06-24 10:54:57.697 [debu]  coderd: GET  host=localhost:45103  path=/api/v2/users/crazy-montalcini7-lk4  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.697257345Z"  requestor_id=402d5ce9-49bc-4df5-a39d-35ffb6d3cbb8  requestor_name=testuser  requestor_email=""  took=634.992µs  status_code=200  latency_ms=0  params_user=crazy-montalcini7-lk4  request_id=a85156d8-3406-45f1-becc-1080b4ccc77e
    exp_mcp_test.go:784: inserted deleted subagent antagonist tender_payne361 (3188f4bd-9383-48e9-a4b5-3ec5a98051fb) for workspace agent dev (b99b15e9-b866-4036-a729-993d9a1934c8)
    clitest.go:75: invoking command: coder --global-config /tmp/TestExpMcpReporterOK3666211288/002 exp mcp server --agent-url http://localhost:45103 --agent-token 22c32b43-224c-467c-a261-258121d90817 --app-status-slug vscode --ai-agentapi-url http://127.0.0.1:39289 --allowed-tools=coder_report_task
    exp_mcp_test.go:872: 2025-06-24 10:54:57.871: cmd: stdin: "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"initialize\"}\r"
    ptytest.go:132: 2025-06-24 10:54:57.873: cmd: "URL            : Not configured\r"
    ptytest.go:132: 2025-06-24 10:54:57.874: cmd: "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"initialize\"}"
    exp_mcp_test.go:873: 2025-06-24 10:54:57.876: cmd: matched newline = "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"initialize\"}"
    ptytest.go:132: 2025-06-24 10:54:57.882: cmd: "Authentication : None\r"
    ptytest.go:132: 2025-06-24 10:54:57.882: cmd: "Agent URL      : http://localhost:45103\r"
    ptytest.go:132: 2025-06-24 10:54:57.882: cmd: "Task reporter  : Enabled\r"
    ptytest.go:132: 2025-06-24 10:54:57.882: cmd: "AI AgentAPI URL  : http://127.0.0.1:39289\r"
    ptytest.go:132: 2025-06-24 10:54:57.882: cmd: "Screen events  : Enabled\r"
    ptytest.go:132: 2025-06-24 10:54:57.882: cmd: "Failed to watch screen events\r"
    ptytest.go:132: 2025-06-24 10:54:57.883: cmd: "Starting MCP server\r"
    ptytest.go:132: 2025-06-24 10:54:57.883: cmd: "Instructions   : \"\"\r"
    ptytest.go:132: 2025-06-24 10:54:57.883: cmd: "Allowed Tools  : [coder_report_task]\r"
    ptytest.go:132: 2025-06-24 10:54:57.883: cmd: "Press Ctrl+C to stop the server\r"
    ptytest.go:132: 2025-06-24 10:54:57.884: cmd: "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"protocolVersion\":\"2025-03-26\",\"capabilities\":{\"tools\":{\"listChanged\":true}},\"serverInfo\":{\"name\":\"Coder Agent\",\"version\":\"v0.0.0-devel\"}}}"
    exp_mcp_test.go:874: 2025-06-24 10:54:57.885: cmd: matched newline = "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"protocolVersion\":\"2025-03-26\",\"capabilities\":{\"tools\":{\"listChanged\":true}},\"serverInfo\":{\"name\":\"Coder Agent\",\"version\":\"v0.0.0-devel\"}}}"
    exp_mcp_test.go:964: 2025-06-24 10:54:57.885: cmd: stdin: "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\", \"params\": {\"name\": \"coder_report_task\", \"arguments\": {\"state\": \"working\", \"summary\": \"doing work\", \"link\": \"https://dev.coder.com\"}}}\r"
    ptytest.go:132: 2025-06-24 10:54:57.886: cmd: "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\", \"params\": {\"name\": \"coder_report_task\", \"arguments\": {\"state\": \"working\", \"summary\": \"doing work\", \"link\": \"https://dev.coder.com\"}}}"
    ptytest.go:132: 2025-06-24 10:54:57.886: cmd: "{\"jsonrpc\":\"2.0\",\"id\":3,\"result\":{\"content\":[{\"type\":\"text\",\"text\":\"{\\\"message\\\":\\\"Thanks for reporting!\\\"}\\n\"}]}}"
    exp_mcp_test.go:965: 2025-06-24 10:54:57.888: cmd: matched newline = "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\", \"params\": {\"name\": \"coder_report_task\", \"arguments\": {\"state\": \"working\", \"summary\": \"doing work\", \"link\": \"https://dev.coder.com\"}}}"
    exp_mcp_test.go:966: 2025-06-24 10:54:57.893: cmd: matched newline = "{\"jsonrpc\":\"2.0\",\"id\":3,\"result\":{\"content\":[{\"type\":\"text\",\"text\":\"{\\\"message\\\":\\\"Thanks for reporting!\\\"}\\n\"}]}}"
    t.go:106: 2025-06-24 10:54:57.900 [debu]  coderd: PATCH  host=localhost:45103  path=/api/v2/workspaceagents/me/app-status  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.888880232Z"  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=11.745746ms  status_code=200  latency_ms=11  request_id=4ceec621-b35f-4f45-bc19-48256298c4ec
    t.go:106: 2025-06-24 10:54:57.973 [debu]  coderd: PATCH  host=localhost:45103  path=/api/v2/workspaceagents/me/app-status  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.901396102Z"  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=71.693011ms  status_code=200  latency_ms=71  request_id=dbd2a769-c8c9-465b-85d7-a9f38d79f406
    exp_mcp_test.go:964: 2025-06-24 10:54:57.976: cmd: stdin: "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\", \"params\": {\"name\": \"coder_report_task\", \"arguments\": {\"state\": \"failure\", \"summary\": \"oops\", \"link\": \"\"}}}\r"
    ptytest.go:132: 2025-06-24 10:54:57.976: cmd: "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\", \"params\": {\"name\": \"coder_report_task\", \"arguments\": {\"state\": \"failure\", \"summary\": \"oops\", \"link\": \"\"}}}"
    ptytest.go:132: 2025-06-24 10:54:57.976: cmd: "{\"jsonrpc\":\"2.0\",\"id\":3,\"result\":{\"content\":[{\"type\":\"text\",\"text\":\"{\\\"message\\\":\\\"Thanks for reporting!\\\"}\\n\"}]}}"
    exp_mcp_test.go:965: 2025-06-24 10:54:57.976: cmd: matched newline = "{\"jsonrpc\":\"2.0\",\"id\":3,\"method\":\"tools/call\", \"params\": {\"name\": \"coder_report_task\", \"arguments\": {\"state\": \"failure\", \"summary\": \"oops\", \"link\": \"\"}}}"
    exp_mcp_test.go:966: 2025-06-24 10:54:57.976: cmd: matched newline = "{\"jsonrpc\":\"2.0\",\"id\":3,\"result\":{\"content\":[{\"type\":\"text\",\"text\":\"{\\\"message\\\":\\\"Thanks for reporting!\\\"}\\n\"}]}}"
    t.go:106: 2025-06-24 10:54:57.998 [debu]  coderd: PATCH  host=localhost:45103  path=/api/v2/workspaceagents/me/app-status  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.97712565Z"  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=21.028007ms  status_code=200  latency_ms=21  request_id=05a3d725-eee3-471d-99c1-e53063cfcf6b
    t.go:106: 2025-06-24 10:54:58.069 [debu]  coderd: PATCH  host=localhost:45103  path=/api/v2/workspaceagents/me/app-status  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.999199035Z"  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=70.066354ms  status_code=200  latency_ms=70  request_id=b63392b4-0267-4500-8602-04ed8339f44b
    exp_mcp_test.go:974:
                Error Trace:    /home/coder/repos/coder/cli/exp_mcp_test.go:974
                Error:          Not equal:
                                expected: "idle"
                                actual  : "working"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1,2 +1,2 @@
                                -(codersdk.WorkspaceAppStatusState) (len=4) "idle"
                                +(codersdk.WorkspaceAppStatusState) (len=7) "working"

                Test:           TestExpMcpReporter/OK
    ptytest.go:96: 2025-06-24 10:54:58.069: cmd: closing expecter: PTY close
    ptytest.go:121: 2025-06-24 10:54:58.071: cmd: copy done: <nil>
    ptytest.go:122: 2025-06-24 10:54:58.071: cmd: closing out
    ptytest.go:124: 2025-06-24 10:54:58.071: cmd: closed out: <nil>
    ptytest.go:87: 2025-06-24 10:54:58.071: cmd: closing logw
    ptytest.go:89: 2025-06-24 10:54:58.071: cmd: closed logw: <nil>
    ptytest.go:87: 2025-06-24 10:54:58.071: cmd: closing logr
    ptytest.go:89: 2025-06-24 10:54:58.071: cmd: closed logr: <nil>
    ptytest.go:113: 2025-06-24 10:54:58.071: cmd: closed expecter
    exp_mcp_test.go:867:
                Error Trace:    /home/coder/repos/coder/cli/exp_mcp_test.go:867
                                                        /usr/local/go/src/runtime/asm_amd64.s:1700
                Error:          Received unexpected error:
                                running command "coder exp mcp server": read /dev/pts/2: file already closed
                Test:           TestExpMcpReporter/OK
    ptytest.go:96: 2025-06-24 10:54:58.071: cmd: closing expecter: PTY close
    ptytest.go:121: 2025-06-24 10:54:58.073: cmd: copy done: <nil>
    ptytest.go:122: 2025-06-24 10:54:58.073: cmd: closing out
    ptytest.go:124: 2025-06-24 10:54:58.073: cmd: closed out: <nil>
    ptytest.go:87: 2025-06-24 10:54:58.073: cmd: closing logw
    ptytest.go:89: 2025-06-24 10:54:58.073: cmd: closed logw: <nil>
    ptytest.go:87: 2025-06-24 10:54:58.073: cmd: closing logr
    ptytest.go:89: 2025-06-24 10:54:58.073: cmd: closed logr: <nil>
    ptytest.go:113: 2025-06-24 10:54:58.073: cmd: closed expecter
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: failed to receive DERP message  error="context canceled"
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: main context canceled; do graceful disconnect
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: sent disconnect
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.coord: peerReadLoop got request  peer_id=eb5f0702-71b0-4bc2-821c-6ce9bc88a40f  peer_name=inmemoryeb5f0702-71b0-4bc2-821c-6ce9bc88a40f
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.coord: got tunnel peers  peer_id=eb5f0702-71b0-4bc2-821c-6ce9bc88a40f  tunnel_peers=[]
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.workspaceapps.stats_collector: flushing workspace app stats
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: failed to read from protocol  error=EOF
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.workspaceapps.stats_collector: flushed workspace app stats
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: responses closed after disconnect
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.workspaceapps.stats_collector: workspace app stats collector stopped
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd: polling DERPMap exited
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.coord: peerReadLoop channel closed  peer_id=eb5f0702-71b0-4bc2-821c-6ce9bc88a40f  peer_name=inmemoryeb5f0702-71b0-4bc2-821c-6ce9bc88a40f
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd: PATCH  host=localhost:45103  path=/api/v2/workspaceagents/me/app-status  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:58.071185913Z"  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=3.152489ms  status_code=200  latency_ms=3  request_id=b3e59340-97c5-47ff-8c72-4962a9841815
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.coord: lostPeer  peer_id=eb5f0702-71b0-4bc2-821c-6ce9bc88a40f
    t.go:106: 2025-06-24 10:54:58.074 [info]  coderd.servertailnet: closing server tailnet
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd: GET  host=localhost:45103  path=/api/v2/workspaces/a468801b-e002-4d5c-b17b-66bb2e406656/watch  proto=HTTP/1.1  remote_addr=127.0.0.1  start="2025-06-24T10:54:57.751483748Z"  workspace_name=strange_stonebraker360  requestor_id=c1c28ad1-9f53-42e0-866e-6b7a20b10e83  requestor_name=crazy-montalcini7-lk4  requestor_email=""  took=322.984455ms  status_code=200  latency_ms=322  params_workspace=a468801b-e002-4d5c-b17b-66bb2e406656  request_id=3d4e1007-910b-44b6-bc9c-ee49d62008af
    t.go:106: 2025-06-24 10:54:58.074 [info]  coderd.servertailnet: closing tailnet Conn
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: closing configMaps configLoop
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: closing nodeUpdater updateLoop
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: stopped expiring old agents
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: closed netstack
    t.go:106: 2025-06-24 10:54:58.074 [debu]  coderd.servertailnet: tailnet API v2+ connection lost
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Routine: receive incoming mkReceiveFunc - stopped
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Routine: receive incoming mkReceiveFunc - stopped
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet.net.wgengine: [v1] warning: fakeRouter.Close: not implemented.
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Device closing
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Routine: receive incoming receiveDERP - stopped
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet.net.wgengine: wg: [v2] Device closed
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.servertailnet: server tailnet close complete
    t.go:106: 2025-06-24 10:54:58.075 [debu]  coderd.workspace_usage_tracker: workspace usage tracker loop exited
    stuntest.go:63: STUN server shutdown
    coderdtest.go:444: closing coderdtest server on 127.0.0.1:45103
    coderdtest.go:446: closed coderdtest server on 127.0.0.1:45103
    --- FAIL: TestExpMcpReporter/OK (0.69s)
```

--- a/cli/exp_mcp.go
+++ b/cli/exp_mcp.go
@@ -362,11 +362,19 @@ func (*RootCmd) mcpConfigureCursor() *serpent.Command {
 }
 
 type taskReport struct {
-	link         string
-	messageID    int64
+	// link is optional.
+	link string
+	// messageID must be set if this update is from a *user* message. A user
+	// message only happens when interacting via the AI AgentAPI (as opposed to
+	// interacting with the terminal directly).
+	messageID *int64
+	// selfReported must be set if the update is directly from the AI agent
+	// (as opposed to the screen watcher).
 	selfReported bool
-	state        codersdk.WorkspaceAppStatusState
-	summary      string
+	// state must always be set.
+	state codersdk.WorkspaceAppStatusState
+	// summary is optional.
+	summary string
 }
 
 type mcpServer struct {
@@ -388,17 +396,22 @@ func (r *RootCmd) mcpServer() *serpent.Command {
 	return &serpent.Command{
 		Use: "server",
 		Handler: func(inv *serpent.Invocation) error {
-			// lastUserMessageID is the ID of the last *user* message that we saw.  A
-			// user message only happens when interacting via the AI AgentAPI (as
-			// opposed to interacting with the terminal directly).
-			var lastUserMessageID int64
 			var lastReport taskReport
 			// Create a queue that skips duplicates and preserves summaries.
 			queue := cliutil.NewQueue[taskReport](512).WithPredicate(func(report taskReport) (taskReport, bool) {
-				// Use "working" status if this is a new user message.  If this is not a
-				// new user message, and the status is "working" and not self-reported
-				// (meaning it came from the screen watcher), then it means one of two
-				// things:
+				// Avoid queuing empty statuses (this would probably indicate a
+				// developer error)
+				if report.state == "" {
+					return report, false
+				}
+				// If this is a user message, discard if it is not new.
+				if report.messageID != nil && lastReport.messageID != nil &&
+					*lastReport.messageID >= *report.messageID {
+					return report, false
+				}
+				// If this is not a user message, and the status is "working" and not
+				// self-reported (meaning it came from the screen watcher), then it
+				// means one of two things:
 				//
 				// 1. The AI agent is not working; the user is interacting with the
 				//    terminal directly.
@@ -415,10 +428,15 @@ func (r *RootCmd) mcpServer() *serpent.Command {
 				// user manually submits a new prompt and the AI agent becomes active
 				// (and does not update itself), but it avoids spamming useless status
 				// updates as the user is typing, so the tradeoff is worth it.
-				if report.messageID > lastUserMessageID {
-					report.state = codersdk.WorkspaceAppStatusStateWorking
-				} else if report.state == codersdk.WorkspaceAppStatusStateWorking && !report.selfReported && lastReport.state != "" {
+				if report.messageID == nil &&
+					report.state == codersdk.WorkspaceAppStatusStateWorking &&
+					!report.selfReported && lastReport.state != "" {
 					return report, false
+				}
+				// Keep track of the last message ID so we can tell when a message is
+				// new or if it has been re-emitted.
+				if report.messageID == nil {
+					report.messageID = lastReport.messageID
 				}
 				// Preserve previous message and URI if there was no message.
 				if report.summary == "" {
@@ -607,7 +625,8 @@ func (s *mcpServer) startWatcher(ctx context.Context, inv *serpent.Invocation) {
 				case agentapi.EventMessageUpdate:
 					if ev.Role == agentapi.RoleUser {
 						err := s.queue.Push(taskReport{
-							messageID: ev.Id,
+							messageID: &ev.Id,
+							state:     codersdk.WorkspaceAppStatusStateWorking,
 						})
 						if err != nil {
 							cliui.Warnf(inv.Stderr, "Failed to queue update: %s", err)

--- a/cli/exp_mcp.go
+++ b/cli/exp_mcp.go
@@ -676,14 +676,19 @@ func (s *mcpServer) startServer(ctx context.Context, inv *serpent.Invocation, in
 	// Add tool dependencies.
 	toolOpts := []func(*toolsdk.Deps){
 		toolsdk.WithTaskReporter(func(args toolsdk.ReportTaskArgs) error {
+			// The agent does not reliably report its status correctly.  If AgentAPI
+			// is enabled, we will always set the status to "working" when we get an
+			// MCP message, and rely on the screen watcher to eventually catch the
+			// idle state.
+			state := codersdk.WorkspaceAppStatusStateWorking
+			if s.aiAgentAPIClient == nil {
+				state = codersdk.WorkspaceAppStatusState(args.State)
+			}
 			return s.queue.Push(taskReport{
 				link:         args.Link,
 				selfReported: true,
-				// The agent does not reliably report its status correctly.  We will
-				// always set the status to "working" when we get an MCP message, and
-				// rely on the screen watcher to eventually catch the idle state.
-				state:   codersdk.WorkspaceAppStatusStateWorking,
-				summary: args.Summary,
+				state:        state,
+				summary:      args.Summary,
 			})
 		}),
 	}

--- a/cli/exp_mcp.go
+++ b/cli/exp_mcp.go
@@ -679,8 +679,11 @@ func (s *mcpServer) startServer(ctx context.Context, inv *serpent.Invocation, in
 			return s.queue.Push(taskReport{
 				link:         args.Link,
 				selfReported: true,
-				state:        codersdk.WorkspaceAppStatusState(args.State),
-				summary:      args.Summary,
+				// The agent does not reliably report its status correctly.  We will
+				// always set the status to "working" when we get an MCP message, and
+				// rely on the screen watcher to eventually catch the idle state.
+				state:   codersdk.WorkspaceAppStatusStateWorking,
+				summary: args.Summary,
 			})
 		}),
 	}

--- a/cli/exp_mcp.go
+++ b/cli/exp_mcp.go
@@ -399,18 +399,25 @@ func (r *RootCmd) mcpServer() *serpent.Command {
 				// new user message, and the status is "working" and not self-reported
 				// (meaning it came from the screen watcher), then it means one of two
 				// things:
-				// 1. The AI agent is still working, so there is nothing to update.
-				// 2. The AI agent stopped working, then the user has interacted with
-				//    the terminal directly.  For now, we are ignoring these updates.
-				//    This risks missing cases where the user manually submits a new
-				//    prompt and the AI agent becomes active and does not update itself,
-				//    but it avoids spamming useless status updates as the user is
-				//    typing, so the tradeoff is worth it.  In the future, if we can
-				//    reliably distinguish between user and AI agent activity, we can
-				//    change this.
+				//
+				// 1. The AI agent is not working; the user is interacting with the
+				//    terminal directly.
+				// 2. The AI agent is working.
+				//
+				// At the moment, we have no way to tell the difference between these
+				// two states. In the future, if we can reliably distinguish between
+				// user and AI agent activity, we can change this.
+				//
+				// If this is our first update, we assume it is the AI agent working and
+				// accept the update.
+				//
+				// Otherwise we discard the update.  This risks missing cases where the
+				// user manually submits a new prompt and the AI agent becomes active
+				// (and does not update itself), but it avoids spamming useless status
+				// updates as the user is typing, so the tradeoff is worth it.
 				if report.messageID > lastUserMessageID {
 					report.state = codersdk.WorkspaceAppStatusStateWorking
-				} else if report.state == codersdk.WorkspaceAppStatusStateWorking && !report.selfReported {
+				} else if report.state == codersdk.WorkspaceAppStatusStateWorking && !report.selfReported && lastReport.state != "" {
 					return report, false
 				}
 				// Preserve previous message and URI if there was no message.

--- a/cli/exp_mcp_test.go
+++ b/cli/exp_mcp_test.go
@@ -835,7 +835,7 @@ func TestExpMcpReporter(t *testing.T) {
 				},
 				// Agent messages are ignored.
 				{
-					event: makeMessageEvent(1, agentapi.RoleAgent),
+					event: makeMessageEvent(0, agentapi.RoleAgent),
 				},
 				// AI agent reports that it failed and URI is blank.
 				{
@@ -854,7 +854,7 @@ func TestExpMcpReporter(t *testing.T) {
 				// ... but this time we have a new user message so we know there is AI
 				// agent activity.  This time the "working" update will not be skipped.
 				{
-					event: makeMessageEvent(2, agentapi.RoleUser),
+					event: makeMessageEvent(1, agentapi.RoleUser),
 					expected: &codersdk.WorkspaceAppStatus{
 						State:   codersdk.WorkspaceAppStatusStateWorking,
 						Message: "oops",
@@ -891,6 +891,33 @@ func TestExpMcpReporter(t *testing.T) {
 					event: makeStatusEvent(agentapi.StatusStable),
 					expected: &codersdk.WorkspaceAppStatus{
 						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "",
+						URI:     "",
+					},
+				},
+				// Zero ID should be accepted.
+				{
+					event: makeMessageEvent(0, agentapi.RoleUser),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "",
+						URI:     "",
+					},
+				},
+				// Stable again.
+				{
+					event: makeStatusEvent(agentapi.StatusStable),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "",
+						URI:     "",
+					},
+				},
+				// Next ID.
+				{
+					event: makeMessageEvent(1, agentapi.RoleUser),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
 						Message: "",
 						URI:     "",
 					},
@@ -954,6 +981,7 @@ func TestExpMcpReporter(t *testing.T) {
 					case w, ok := <-watcher:
 						require.True(t, ok, "watch channel closed")
 						if w.LatestAppStatus != nil && w.LatestAppStatus.ID != lastAppStatus.ID {
+							t.Logf("Got status update: %s > %s", lastAppStatus.State, w.LatestAppStatus.State)
 							lastAppStatus = *w.LatestAppStatus
 							return lastAppStatus
 						}

--- a/cli/exp_mcp_test.go
+++ b/cli/exp_mcp_test.go
@@ -793,8 +793,9 @@ func TestExpMcpReporter(t *testing.T) {
 	}
 
 	runs := []struct {
-		name  string
-		tests []test
+		name            string
+		tests           []test
+		disableAgentAPI bool
 	}{
 		// In this run the AI agent starts with a state change but forgets to update
 		// that it finished.
@@ -954,6 +955,29 @@ func TestExpMcpReporter(t *testing.T) {
 				},
 			},
 		},
+		// When AgentAPI is not being used, we accept agent state updates as-is.
+		{
+			name: "KeepAgentState",
+			tests: []test{
+				{
+					state:   codersdk.WorkspaceAppStatusStateWorking,
+					summary: "doing work",
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "doing work",
+					},
+				},
+				{
+					state:   codersdk.WorkspaceAppStatusStateIdle,
+					summary: "finished",
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "finished",
+					},
+				},
+			},
+			disableAgentAPI: true,
+		},
 	}
 
 	for _, run := range runs {
@@ -980,25 +1004,6 @@ func TestExpMcpReporter(t *testing.T) {
 				return a
 			}).Do()
 
-			// Mock the AI AgentAPI server.
-			listening := make(chan func(sse codersdk.ServerSentEvent) error)
-			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				send, closed, err := httpapi.ServerSentEventSender(w, r)
-				if err != nil {
-					httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
-						Message: "Internal error setting up server-sent events.",
-						Detail:  err.Error(),
-					})
-					return
-				}
-				// Send initial message.
-				send(*makeMessageEvent(0, agentapi.RoleAgent))
-				listening <- send
-				<-closed
-			}))
-			t.Cleanup(srv.Close)
-			aiAgentAPIURL := srv.URL
-
 			// Watch the workspace for changes.
 			watcher, err := client.WatchWorkspace(ctx, r.Workspace.ID)
 			require.NoError(t, err)
@@ -1019,15 +1024,39 @@ func TestExpMcpReporter(t *testing.T) {
 				}
 			}
 
-			inv, _ := clitest.New(t,
+			args := []string{
 				"exp", "mcp", "server",
-				// We need the agent credentials, AI AgentAPI url, and a slug for reporting.
+				// We need the agent credentials, AI AgentAPI url (if not
+				// disabled), and a slug for reporting.
 				"--agent-url", client.URL.String(),
 				"--agent-token", r.AgentToken,
 				"--app-status-slug", "vscode",
-				"--ai-agentapi-url", aiAgentAPIURL,
 				"--allowed-tools=coder_report_task",
-			)
+			}
+
+			// Mock the AI AgentAPI server.
+			listening := make(chan func(sse codersdk.ServerSentEvent) error)
+			if !run.disableAgentAPI {
+				srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					send, closed, err := httpapi.ServerSentEventSender(w, r)
+					if err != nil {
+						httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
+							Message: "Internal error setting up server-sent events.",
+							Detail:  err.Error(),
+						})
+						return
+					}
+					// Send initial message.
+					send(*makeMessageEvent(0, agentapi.RoleAgent))
+					listening <- send
+					<-closed
+				}))
+				t.Cleanup(srv.Close)
+				aiAgentAPIURL := srv.URL
+				args = append(args, "--ai-agentapi-url", aiAgentAPIURL)
+			}
+
+			inv, _ := clitest.New(t, args...)
 			inv = inv.WithContext(ctx)
 
 			pty := ptytest.New(t)
@@ -1050,7 +1079,10 @@ func TestExpMcpReporter(t *testing.T) {
 			_ = pty.ReadLine(ctx) // ignore echo
 			_ = pty.ReadLine(ctx) // ignore init response
 
-			sender := <-listening
+			var sender func(sse codersdk.ServerSentEvent) error
+			if !run.disableAgentAPI {
+				sender = <-listening
+			}
 
 			for _, test := range run.tests {
 				if test.event != nil {

--- a/cli/exp_mcp_test.go
+++ b/cli/exp_mcp_test.go
@@ -822,7 +822,8 @@ func TestExpMcpReporter(t *testing.T) {
 						URI:     "https://dev.coder.com",
 					},
 				},
-				// A completed update at this point from the watcher should be discarded.
+				// A stable update now from the watcher should be discarded, as it is a
+				// duplicate.
 				{
 					event: makeStatusEvent(agentapi.StatusStable),
 				},
@@ -830,22 +831,17 @@ func TestExpMcpReporter(t *testing.T) {
 				// new user message.  This could be the AI agent being active again, but
 				// it could also be the user messing around.  We will prefer not updating
 				// the status so the "working" update here should be skipped.
+				//
+				// TODO: How do we test the no-op updates?  This update is skipped
+				// because of the logic mentioned above, but how do we prove this update
+				// was skipped because of that and not that the next update was skipped
+				// because it is a duplicate state?  We could mock the queue?
 				{
 					event: makeStatusEvent(agentapi.StatusRunning),
 				},
 				// Agent messages are ignored.
 				{
 					event: makeMessageEvent(0, agentapi.RoleAgent),
-				},
-				// AI agent reports that it failed and URI is blank.
-				{
-					state:   codersdk.WorkspaceAppStatusStateFailure,
-					summary: "oops",
-					expected: &codersdk.WorkspaceAppStatus{
-						State:   codersdk.WorkspaceAppStatusStateFailure,
-						Message: "oops",
-						URI:     "",
-					},
 				},
 				// The watcher reports the screen is active again...
 				{
@@ -857,8 +853,8 @@ func TestExpMcpReporter(t *testing.T) {
 					event: makeMessageEvent(1, agentapi.RoleUser),
 					expected: &codersdk.WorkspaceAppStatus{
 						State:   codersdk.WorkspaceAppStatusStateWorking,
-						Message: "oops",
-						URI:     "",
+						Message: "doing work",
+						URI:     "https://dev.coder.com",
 					},
 				},
 				// Watcher reports stable again.
@@ -866,8 +862,8 @@ func TestExpMcpReporter(t *testing.T) {
 					event: makeStatusEvent(agentapi.StatusStable),
 					expected: &codersdk.WorkspaceAppStatus{
 						State:   codersdk.WorkspaceAppStatusStateIdle,
-						Message: "oops",
-						URI:     "",
+						Message: "doing work",
+						URI:     "https://dev.coder.com",
 					},
 				},
 			},
@@ -920,6 +916,40 @@ func TestExpMcpReporter(t *testing.T) {
 						State:   codersdk.WorkspaceAppStatusStateWorking,
 						Message: "",
 						URI:     "",
+					},
+				},
+			},
+		},
+		// We ignore the state from the agent and assume "working".
+		{
+			name: "IgnoreAgentState",
+			// AI agent reports that it is finished but the summary says it is doing
+			// work.
+			tests: []test{
+				{
+					state:   codersdk.WorkspaceAppStatusStateIdle,
+					summary: "doing work",
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "doing work",
+					},
+				},
+				// AI agent reports finished again, with a matching summary.  We still
+				// assume it is working.
+				{
+					state:   codersdk.WorkspaceAppStatusStateIdle,
+					summary: "finished",
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "finished",
+					},
+				},
+				// Once the watcher reports stable, then we record idle.
+				{
+					event: makeStatusEvent(agentapi.StatusStable),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "finished",
 					},
 				},
 			},

--- a/cli/exp_mcp_test.go
+++ b/cli/exp_mcp_test.go
@@ -763,220 +763,261 @@ func TestExpMcpReporter(t *testing.T) {
 		<-cmdDone
 	})
 
-	t.Run("OK", func(t *testing.T) {
-		t.Parallel()
+	makeStatusEvent := func(status agentapi.AgentStatus) *codersdk.ServerSentEvent {
+		return &codersdk.ServerSentEvent{
+			Type: ServerSentEventTypeStatusChange,
+			Data: agentapi.EventStatusChange{
+				Status: status,
+			},
+		}
+	}
 
-		// Create a test deployment and workspace.
-		client, db := coderdtest.NewWithDatabase(t, nil)
-		user := coderdtest.CreateFirstUser(t, client)
-		client, user2 := coderdtest.CreateAnotherUser(t, client, user.OrganizationID)
+	makeMessageEvent := func(id int64, role agentapi.ConversationRole) *codersdk.ServerSentEvent {
+		return &codersdk.ServerSentEvent{
+			Type: ServerSentEventTypeMessageUpdate,
+			Data: agentapi.EventMessageUpdate{
+				Id:   id,
+				Role: role,
+			},
+		}
+	}
 
-		r := dbfake.WorkspaceBuild(t, db, database.WorkspaceTable{
-			OrganizationID: user.OrganizationID,
-			OwnerID:        user2.ID,
-		}).WithAgent(func(a []*proto.Agent) []*proto.Agent {
-			a[0].Apps = []*proto.App{
+	type test struct {
+		// event simulates an event from the screen watcher.
+		event *codersdk.ServerSentEvent
+		// state, summary, and uri simulate a tool call from the AI agent.
+		state    codersdk.WorkspaceAppStatusState
+		summary  string
+		uri      string
+		expected *codersdk.WorkspaceAppStatus
+	}
+
+	runs := []struct {
+		name  string
+		tests []test
+	}{
+		// In this run the AI agent starts with a state change but forgets to update
+		// that it finished.
+		{
+			name: "Active",
+			tests: []test{
+				// First the AI agent updates with a state change.
 				{
-					Slug: "vscode",
+					state:   codersdk.WorkspaceAppStatusStateWorking,
+					summary: "doing work",
+					uri:     "https://dev.coder.com",
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "doing work",
+						URI:     "https://dev.coder.com",
+					},
 				},
-			}
-			return a
-		}).Do()
-
-		makeStatusEvent := func(status agentapi.AgentStatus) *codersdk.ServerSentEvent {
-			return &codersdk.ServerSentEvent{
-				Type: ServerSentEventTypeStatusChange,
-				Data: agentapi.EventStatusChange{
-					Status: status,
+				// Terminal goes quiet but the AI agent forgot the update, and it is
+				// caught by the screen watcher.  Message and URI are preserved.
+				{
+					event: makeStatusEvent(agentapi.StatusStable),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "doing work",
+						URI:     "https://dev.coder.com",
+					},
 				},
-			}
-		}
-
-		makeMessageEvent := func(id int64, role agentapi.ConversationRole) *codersdk.ServerSentEvent {
-			return &codersdk.ServerSentEvent{
-				Type: ServerSentEventTypeMessageUpdate,
-				Data: agentapi.EventMessageUpdate{
-					Id:   id,
-					Role: role,
+				// A completed update at this point from the watcher should be discarded.
+				{
+					event: makeStatusEvent(agentapi.StatusStable),
 				},
-			}
-		}
+				// Terminal becomes active again according to the screen watcher, but no
+				// new user message.  This could be the AI agent being active again, but
+				// it could also be the user messing around.  We will prefer not updating
+				// the status so the "working" update here should be skipped.
+				{
+					event: makeStatusEvent(agentapi.StatusRunning),
+				},
+				// Agent messages are ignored.
+				{
+					event: makeMessageEvent(1, agentapi.RoleAgent),
+				},
+				// AI agent reports that it failed and URI is blank.
+				{
+					state:   codersdk.WorkspaceAppStatusStateFailure,
+					summary: "oops",
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateFailure,
+						Message: "oops",
+						URI:     "",
+					},
+				},
+				// The watcher reports the screen is active again...
+				{
+					event: makeStatusEvent(agentapi.StatusRunning),
+				},
+				// ... but this time we have a new user message so we know there is AI
+				// agent activity.  This time the "working" update will not be skipped.
+				{
+					event: makeMessageEvent(2, agentapi.RoleUser),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "oops",
+						URI:     "",
+					},
+				},
+				// Watcher reports stable again.
+				{
+					event: makeStatusEvent(agentapi.StatusStable),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "oops",
+						URI:     "",
+					},
+				},
+			},
+		},
+		// In this run the AI agent never sends any state changes.
+		{
+			name: "Inactive",
+			tests: []test{
+				// The "working" status from the watcher should be accepted, even though
+				// there is no new user message, because it is the first update.
+				{
+					event: makeStatusEvent(agentapi.StatusRunning),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateWorking,
+						Message: "",
+						URI:     "",
+					},
+				},
+				// Stable update should be accepted.
+				{
+					event: makeStatusEvent(agentapi.StatusStable),
+					expected: &codersdk.WorkspaceAppStatus{
+						State:   codersdk.WorkspaceAppStatusStateIdle,
+						Message: "",
+						URI:     "",
+					},
+				},
+			},
+		},
+	}
 
-		ctx, cancel := context.WithCancel(testutil.Context(t, testutil.WaitShort))
+	for _, run := range runs {
+		run := run
+		t.Run(run.name, func(t *testing.T) {
+			t.Parallel()
 
-		// Mock the AI AgentAPI server.
-		listening := make(chan func(sse codersdk.ServerSentEvent) error)
-		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			send, closed, err := httpapi.ServerSentEventSender(w, r)
-			if err != nil {
-				httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
-					Message: "Internal error setting up server-sent events.",
-					Detail:  err.Error(),
-				})
-				return
-			}
-			// Send initial message.
-			send(*makeMessageEvent(0, agentapi.RoleAgent))
-			listening <- send
-			<-closed
-		}))
-		t.Cleanup(srv.Close)
-		aiAgentAPIURL := srv.URL
+			ctx, cancel := context.WithCancel(testutil.Context(t, testutil.WaitShort))
 
-		// Watch the workspace for changes.
-		watcher, err := client.WatchWorkspace(ctx, r.Workspace.ID)
-		require.NoError(t, err)
-		var lastAppStatus codersdk.WorkspaceAppStatus
-		nextUpdate := func() codersdk.WorkspaceAppStatus {
-			for {
-				select {
-				case <-ctx.Done():
-					require.FailNow(t, "timed out waiting for status update")
-				case w, ok := <-watcher:
-					require.True(t, ok, "watch channel closed")
-					if w.LatestAppStatus != nil && w.LatestAppStatus.ID != lastAppStatus.ID {
-						lastAppStatus = *w.LatestAppStatus
-						return lastAppStatus
+			// Create a test deployment and workspace.
+			client, db := coderdtest.NewWithDatabase(t, nil)
+			user := coderdtest.CreateFirstUser(t, client)
+			client, user2 := coderdtest.CreateAnotherUser(t, client, user.OrganizationID)
+
+			r := dbfake.WorkspaceBuild(t, db, database.WorkspaceTable{
+				OrganizationID: user.OrganizationID,
+				OwnerID:        user2.ID,
+			}).WithAgent(func(a []*proto.Agent) []*proto.Agent {
+				a[0].Apps = []*proto.App{
+					{
+						Slug: "vscode",
+					},
+				}
+				return a
+			}).Do()
+
+			// Mock the AI AgentAPI server.
+			listening := make(chan func(sse codersdk.ServerSentEvent) error)
+			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				send, closed, err := httpapi.ServerSentEventSender(w, r)
+				if err != nil {
+					httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
+						Message: "Internal error setting up server-sent events.",
+						Detail:  err.Error(),
+					})
+					return
+				}
+				// Send initial message.
+				send(*makeMessageEvent(0, agentapi.RoleAgent))
+				listening <- send
+				<-closed
+			}))
+			t.Cleanup(srv.Close)
+			aiAgentAPIURL := srv.URL
+
+			// Watch the workspace for changes.
+			watcher, err := client.WatchWorkspace(ctx, r.Workspace.ID)
+			require.NoError(t, err)
+			var lastAppStatus codersdk.WorkspaceAppStatus
+			nextUpdate := func() codersdk.WorkspaceAppStatus {
+				for {
+					select {
+					case <-ctx.Done():
+						require.FailNow(t, "timed out waiting for status update")
+					case w, ok := <-watcher:
+						require.True(t, ok, "watch channel closed")
+						if w.LatestAppStatus != nil && w.LatestAppStatus.ID != lastAppStatus.ID {
+							lastAppStatus = *w.LatestAppStatus
+							return lastAppStatus
+						}
 					}
 				}
 			}
-		}
 
-		inv, _ := clitest.New(t,
-			"exp", "mcp", "server",
-			// We need the agent credentials, AI AgentAPI url, and a slug for reporting.
-			"--agent-url", client.URL.String(),
-			"--agent-token", r.AgentToken,
-			"--app-status-slug", "vscode",
-			"--ai-agentapi-url", aiAgentAPIURL,
-			"--allowed-tools=coder_report_task",
-		)
-		inv = inv.WithContext(ctx)
+			inv, _ := clitest.New(t,
+				"exp", "mcp", "server",
+				// We need the agent credentials, AI AgentAPI url, and a slug for reporting.
+				"--agent-url", client.URL.String(),
+				"--agent-token", r.AgentToken,
+				"--app-status-slug", "vscode",
+				"--ai-agentapi-url", aiAgentAPIURL,
+				"--allowed-tools=coder_report_task",
+			)
+			inv = inv.WithContext(ctx)
 
-		pty := ptytest.New(t)
-		inv.Stdin = pty.Input()
-		inv.Stdout = pty.Output()
-		stderr := ptytest.New(t)
-		inv.Stderr = stderr.Output()
+			pty := ptytest.New(t)
+			inv.Stdin = pty.Input()
+			inv.Stdout = pty.Output()
+			stderr := ptytest.New(t)
+			inv.Stderr = stderr.Output()
 
-		// Run the MCP server.
-		cmdDone := make(chan struct{})
-		go func() {
-			defer close(cmdDone)
-			err := inv.Run()
-			assert.NoError(t, err)
-		}()
+			// Run the MCP server.
+			cmdDone := make(chan struct{})
+			go func() {
+				defer close(cmdDone)
+				err := inv.Run()
+				assert.NoError(t, err)
+			}()
 
-		// Initialize.
-		payload := `{"jsonrpc":"2.0","id":1,"method":"initialize"}`
-		pty.WriteLine(payload)
-		_ = pty.ReadLine(ctx) // ignore echo
-		_ = pty.ReadLine(ctx) // ignore init response
+			// Initialize.
+			payload := `{"jsonrpc":"2.0","id":1,"method":"initialize"}`
+			pty.WriteLine(payload)
+			_ = pty.ReadLine(ctx) // ignore echo
+			_ = pty.ReadLine(ctx) // ignore init response
 
-		sender := <-listening
+			sender := <-listening
 
-		tests := []struct {
-			// event simulates an event from the screen watcher.
-			event *codersdk.ServerSentEvent
-			// state, summary, and uri simulate a tool call from the AI agent.
-			state    codersdk.WorkspaceAppStatusState
-			summary  string
-			uri      string
-			expected *codersdk.WorkspaceAppStatus
-		}{
-			// First the AI agent updates with a state change.
-			{
-				state:   codersdk.WorkspaceAppStatusStateWorking,
-				summary: "doing work",
-				uri:     "https://dev.coder.com",
-				expected: &codersdk.WorkspaceAppStatus{
-					State:   codersdk.WorkspaceAppStatusStateWorking,
-					Message: "doing work",
-					URI:     "https://dev.coder.com",
-				},
-			},
-			// Terminal goes quiet but the AI agent forgot the update, and it is
-			// caught by the screen watcher.  Message and URI are preserved.
-			{
-				event: makeStatusEvent(agentapi.StatusStable),
-				expected: &codersdk.WorkspaceAppStatus{
-					State:   codersdk.WorkspaceAppStatusStateIdle,
-					Message: "doing work",
-					URI:     "https://dev.coder.com",
-				},
-			},
-			// A completed update at this point from the watcher should be discarded.
-			{
-				event: makeStatusEvent(agentapi.StatusStable),
-			},
-			// Terminal becomes active again according to the screen watcher, but no
-			// new user message.  This could be the AI agent being active again, but
-			// it could also be the user messing around.  We will prefer not updating
-			// the status so the "working" update here should be skipped.
-			{
-				event: makeStatusEvent(agentapi.StatusRunning),
-			},
-			// Agent messages are ignored.
-			{
-				event: makeMessageEvent(1, agentapi.RoleAgent),
-			},
-			// AI agent reports that it failed and URI is blank.
-			{
-				state:   codersdk.WorkspaceAppStatusStateFailure,
-				summary: "oops",
-				expected: &codersdk.WorkspaceAppStatus{
-					State:   codersdk.WorkspaceAppStatusStateFailure,
-					Message: "oops",
-					URI:     "",
-				},
-			},
-			// The watcher reports the screen is active again...
-			{
-				event: makeStatusEvent(agentapi.StatusRunning),
-			},
-			// ... but this time we have a new user message so we know there is AI
-			// agent activity.  This time the "working" update will not be skipped.
-			{
-				event: makeMessageEvent(2, agentapi.RoleUser),
-				expected: &codersdk.WorkspaceAppStatus{
-					State:   codersdk.WorkspaceAppStatusStateWorking,
-					Message: "oops",
-					URI:     "",
-				},
-			},
-			// Watcher reports stable again.
-			{
-				event: makeStatusEvent(agentapi.StatusStable),
-				expected: &codersdk.WorkspaceAppStatus{
-					State:   codersdk.WorkspaceAppStatusStateIdle,
-					Message: "oops",
-					URI:     "",
-				},
-			},
-		}
-		for _, test := range tests {
-			if test.event != nil {
-				err := sender(*test.event)
-				require.NoError(t, err)
-			} else {
-				// Call the tool and ensure it works.
-				payload := fmt.Sprintf(`{"jsonrpc":"2.0","id":3,"method":"tools/call", "params": {"name": "coder_report_task", "arguments": {"state": %q, "summary": %q, "link": %q}}}`, test.state, test.summary, test.uri)
-				pty.WriteLine(payload)
-				_ = pty.ReadLine(ctx) // ignore echo
-				output := pty.ReadLine(ctx)
-				require.NotEmpty(t, output, "did not receive a response from coder_report_task")
-				// Ensure it is valid JSON.
-				_, err = json.Marshal(output)
-				require.NoError(t, err, "did not receive valid JSON from coder_report_task")
+			for _, test := range run.tests {
+				if test.event != nil {
+					err := sender(*test.event)
+					require.NoError(t, err)
+				} else {
+					// Call the tool and ensure it works.
+					payload := fmt.Sprintf(`{"jsonrpc":"2.0","id":3,"method":"tools/call", "params": {"name": "coder_report_task", "arguments": {"state": %q, "summary": %q, "link": %q}}}`, test.state, test.summary, test.uri)
+					pty.WriteLine(payload)
+					_ = pty.ReadLine(ctx) // ignore echo
+					output := pty.ReadLine(ctx)
+					require.NotEmpty(t, output, "did not receive a response from coder_report_task")
+					// Ensure it is valid JSON.
+					_, err = json.Marshal(output)
+					require.NoError(t, err, "did not receive valid JSON from coder_report_task")
+				}
+				if test.expected != nil {
+					got := nextUpdate()
+					require.Equal(t, got.State, test.expected.State)
+					require.Equal(t, got.Message, test.expected.Message)
+					require.Equal(t, got.URI, test.expected.URI)
+				}
 			}
-			if test.expected != nil {
-				got := nextUpdate()
-				require.Equal(t, got.State, test.expected.State)
-				require.Equal(t, got.Message, test.expected.Message)
-				require.Equal(t, got.URI, test.expected.URI)
-			}
-		}
-		cancel()
-		<-cmdDone
-	})
+			cancel()
+			<-cmdDone
+		})
+	}
 }

